cfbs/commands.py

```diff
@@ -201,15 +202,96 @@ def local_module_copy(module, counter, max_length):
         f"{counter:03d} {pad_right(name, max_length)} @ local                                    (Copied)"
     )
 
+def _get_path_from_url(url):
+    if not url.startswith(("https://", "ssh://", "git://")):
+        if ("://" in url):
+            return user_error("Unsupported URL protocol in '%s'" % url)
+        else:
+            # It's a path already, just remove trailing slashes (if any).
+            return url.strip("/")
+
+    path = None
+    if url.startswith("ssh://"):
+        match = re.match(r"ssh://(\w+)@(.+)", url)
+        if match is not None:
+            path = match[2]
+    path = path or url[url.index("://") + 3:]
+    path = strip_right(path, ".git")
+    path = path.strip("/")
+
+    return path
+
+def _get_git_repo_commit_sha(repo_path):
+    assert os.path.isdir(os.path.join(repo_path, ".git"))
+
+    with open(os.path.join(repo_path, ".git", "HEAD"), "r") as f:
+        head_ref_info = f.read()
+
+    assert head_ref_info.startswith("ref: ")
+    head_ref = head_ref_info[5:].strip()
+
+    with open(os.path.join(repo_path, ".git", head_ref)) as f:
+        return f.read().strip()
+
+def _clone_index_repo(repo_url):
+    assert repo_url.startswith(("https://", "ssh://", "git://"))
+
+    commit = None
+    if "@" in repo_url and (repo_url.rindex("@") > repo_url.rindex(".")):
+        # commit specified in the url
+        repo_url, commit = repo_url.rsplit("@", 1)
+
+    downloads = os.path.join(cfbs_dir(), "downloads")
+
+    repo_path = _get_path_from_url(repo_url)
+    repo_dir = os.path.join(downloads, repo_path)
+    os.makedirs(repo_dir, exist_ok=True)
+
+    if commit is not None:
+        commit_path = os.path.join(repo_dir, commit)
+        sh("git clone --no-checkout %s %s" % (repo_url, commit_path))
+        sh("cd %s; git checkout %s" % (commit_path, commit))
+    else:
+        master_path = os.path.join(repo_dir, "master")
+        sh("git clone %s %s" % (repo_url, master_path))
+        commit = _get_git_repo_commit_sha(master_path)
+
+        commit_path = os.path.join(repo_dir, commit)
+        if os.path.exists(commit_path):
+            # Already cloned in the commit dir, just remove the 'master' clone
+            sh("rm -rf %s" % master_path)
+        else:
+            sh("mv %s %s" % (master_path, commit_path))
+
+    index_path = os.path.join(commit_path, "cfbs.json")
+    if os.path.exists(index_path):
+        return (index_path, commit)
+    else:
+        user_error("Repository '%s' doesn't contain a valid cfbs.json index file" % repo_url)
```

Comment:
I see you are calling it an index file. That's not what I intended, up until now we've had 2 types of JSON files `cfbs.json` (config file for 1 project), `index.json` (index of all available modules).

BUT, what you did is smart and I think we should probably change this, standardize the format so there is only 1 format, `cfbs.json` and it can have an index inline or external path to index.

No need to do anything here, but let's discuss this some more :)